### PR TITLE
Recycle task tokens for invalid matching tasks

### PR DESCRIPTION
**Pull request by carlydf**
## What changed?
Add a `RecycleToken` function to our custom rate limiters. It seems that most of the rate limiters just wrap the `ClockedRateLimiter`, so let the `ClockedRateLimiter`'s `WaitN` function unblock the waiter if recycle is called, thus allowing us to recycle tokens and give them to any waiters if they exist.

## Why?
Lots of invalid tasks combined with Activity task dispatch rate limiting can cause the actual rate to be noticeably below the maximum rate. This will make the actual rate match the max rps, as long as `time_to_recycle` is reasonably short and `recycle_rate` is reasonable (see discussion in the comments)

## How did you test it?
ClockedRateLimiter unit test.

## Potential risks
The max Activity RPS could break.

### CODE REVIEW

File: common/quotas/clocked_rate_limiter.go [hunk 1]
```diff
@@ -137,6 +139,8 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
 		return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
 	case <-waitExpired:
 		return nil
+	case <-l.recycleCh:
+		return nil
```
Comment (prathyushpv) [inline common/quotas/clocked_rate_limiter.go:143]: There could be waiters waiting for more than one token. But this will unblock them as well. Ideally RecycleToken() should only recycle 1 token. If it is not straightforward to track token count, and is not needed for this use case, we should at least add a warning in the RecycleToken() comment.

File: common/quotas/multi_rate_limiter_impl.go [hunk 2]
```diff
@@ -160,7 +160,7 @@ func (rl *MultiRateLimiterImpl) WaitN(ctx context.Context, numToken int) error {
 	}
 }
 
-// Rate returns the rate per second for this rate limiter
+// Rate returns the minimum rate per second for this rate limiter
 func (rl *MultiRateLimiterImpl) Rate() float64 {
 	result := rl.rateLimiters[0].Rate()
 	for _, rateLimiter := range rl.rateLimiters {
```
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/multi_rate_limiter_impl.go:166]: yeah that would be more readable for sure

File: common/quotas/clocked_rate_limiter.go [hunk 3]
```diff
@@ -137,6 +139,8 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
 		return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
 	case <-waitExpired:
 		return nil
+	case <-l.recycleCh:
```
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go:142]: The reason I didn't cancel the reservation here is because I thought that the reservation we are holding when we receive the `l.recycleCh` message is a reservation that _will_ be successfully used by the waiter, who we are freeing.

Now that you say this, I'm thinking that the ideal thing would be to call `reservation.Cancel` on the reservation that was taken and _not_ used (ie. call `reservation.Cancel` within `RecycleToken` or within `task.finish`). However, those places don't have access to the token.

Since the effect of `reservation.Cancel` is just to restore tokens to the rate limiter, considering that other reservations have already been made, it seems safe to cancel the "successful" reservation as a proxy for cancelling the "failed" reservation that we don't have access to.

TL;DR I've convinced myself that I agree we need to cancel the reservation here

File: common/quotas/clocked_rate_limiter.go [hunk 4]
```diff
@@ -151,3 +155,35 @@ func (l ClockedRateLimiter) SetBurstAt(t time.Time, newBurst int) {
...
+//	time_to_recycle * probability_of_recycle * number_of_attempts_per_second <= 1s
+//
+// Therefore, it is also possible for this strategy to be inaccurate if the delay between taking and
+// successfully using a token is greater than one second.
+//
+// Currently, RecycleToken is called when we take a token to attempt a matching task dispatch and
+// then later find out (usually via RPC to History) that the task should not be dispatched.
+// If history rpc takes 10ms --> 100 opportunities for the token to be used that second --> 99% recycle probability is ok.
+// If recycle probability is 50% --> need at least 2 opportunities for token to be used --> 500ms history rpc time is ok.
+func (l ClockedRateLimiter) RecycleToken() {
```
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go:184]: Yes that's ok

File: common/quotas/clocked_rate_limiter.go (hunk 1 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go]: The matching code where we need to use `RecycleToken` only ever reserves one token (same with most of our rate limiter uses -- calling `WaitN` directly is rare), so I decided to make this safe by only letting the recycle go through if the waiter is waiting on one token. If a waiter is waiting on >1 token, they will not be unblocked.

Added the following to the RecycleToken() comment:
```
// Since we don't know how many tokens were reserved by the process calling recycle, we will only unblock
// new reservations that are for one token (otherwise we could recycle a 1-token-reservation and unblock
// a 100-token-reservation). If all waiting processes are waiting for >1 tokens, this is a no-op.
```

Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go]: One reason why it's not straightforward to track token count is because we don't explicitly know how many tokens were reserved when we call recycle.

In the case of matching at least, we do know that it's always 1, because we only ever take 1 token, so the caller of `RecycleToken()` could pass 1 to the recycle call since it knows from the code that it only took 1 token previously. That could definitely be abused though.

Another solution I considered was to add an atomic uint to the rate limiter struct, and increment it with whatever `n` `RecycleToken()` is called with. Then, when the `recycleCh` receives a signal that a recycle just occurred, the waiter checks if there are enough tokens in the bucket to satisfy the reservation, and if there are, it decrements the atomic and takes the reservation, but if there are not, it does nothing and keeps waiting in the select.

I think because most of our RL reservations use only 1 token, the method I pushed a few minutes ago is simpler to test and reason about, but I'm open to the bucket method if you think that it's better for some reason.

Comment (prathyushpv) [reply common/quotas/clocked_rate_limiter.go]: Thanks! I think that is good enough for now. We can add code to handle multiple tokens when we have a use case for that.

Yeah, I was also thinking how to handle multiple tokens. I was thinking if it is better to add `Wait()` and `Recycle()` methods to `Reservation` interface. `Reservation` object will know the number of tokens to recycle. That way we don't need to pass the token count to `Recycle()`. 

One idea is to pass the number of tokens to recycle to `l.recycleCh` instead of empty struct. Then the receiving waiter could check if that is enough to unblock itself. If not, it could pass the same number back to the channel for other waiters. I am not suggesting to implement any of these. It is fine to implement it when we have a use case for that.

File: common/quotas/clocked_rate_limiter_test.go [hunk 5]
```diff
@@ -151,3 +151,81 @@ func TestClockedRateLimiter_Wait_DeadlineWouldExceed(t *testing.T) {
...
+	rl := quotas.NewClockedRateLimiter(rate.NewLimiter(1, 1), ts)
+	ctx := context.Background()
+
+	// take first token
+	assert.NoError(t, rl.Wait(ctx))
+
+	// wait for next token and report when success
+	waiting := make(chan struct{})
+	done := make(chan struct{})
+	go func() {
```
Comment (prathyushpv) [inline common/quotas/clocked_rate_limiter_test.go:168]: Do we need this outer go func()?

**Review by prathyushpv — APPROVED** [review]

File: common/quotas/clocked_rate_limiter.go (hunk 1 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go]: > I was thinking if it is better to add Wait() and Recycle() methods to Reservation interface. Reservation object will know the number of tokens to recycle. That way we don't need to pass the token count to Recycle().
The problem with that idea is that `Wait` doesn't actually return the reservation to the caller, it just takes the reservation and then returns error or nil. So, the caller of `Wait`, which is the process that knows when to call `Recycle`, doesn't have access to the reservation that it waited on.

File: common/quotas/clocked_rate_limiter_test.go (hunk 5 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter_test.go:168]: yes, because we need to start waiting before we call recycle (otherwise recycle will be a no-op)

File: common/quotas/clocked_rate_limiter_test.go [hunk 6]
```diff
@@ -151,3 +151,81 @@ func TestClockedRateLimiter_Wait_DeadlineWouldExceed(t *testing.T) {
...
+		time.Sleep(time.Millisecond)
+		waiting <- struct{}{}
+	}()
+
+	// once a waiter exists, recycle the token instead of advancing time
+	<-waiting
+	rl.RecycleToken()
+
+	// wait until done so we know assert.NoError was called
+	<-done
```
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/clocked_rate_limiter_test.go:187]: since this is a unit test, I don't have access to Eventually unless I refactor the whole test file to become a testify suite

File: common/quotas/clocked_rate_limiter_test.go (hunk 5 above)
Comment (prathyushpv) [reply common/quotas/clocked_rate_limiter_test.go:168]: time.Sleep(time.Millisecond) will ensure that it will wait before we call recycle right. I think that it will still work if you remove the outer goroutine and `waiting` channel.

File: common/quotas/clocked_rate_limiter_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (prathyushpv) [reply common/quotas/clocked_rate_limiter_test.go:187]: you can still use assert.Eventually()

File: common/quotas/clocked_rate_limiter_test.go (hunk 5 above)
Comment (prathyushpv) [reply common/quotas/clocked_rate_limiter_test.go:168]: I was suggesting this:
```
// wait for next token and report when success
	done := make(chan struct{})
	go func() {
		assert.NoError(t, rl.Wait(ctx))
		done <- struct{}{}
	}()

	// The waiting channel isn't enough by itself to confirm waiting, because
	// the inner go func can be called without the rl.Wait() having finished.
	// Use time.Sleep to allow rl.Wait() to complete before signalling `waiting`.
	// 1-millisecond sleep succeeded >50,000x, whereas 1-nanosecond sleep deadlocks.
	time.Sleep(time.Millisecond)

	// once a waiter exists, recycle the token instead of advancing time
	rl.RecycleToken()

	// wait until done so we know assert.NoError was called
	<-done
```

File: common/quotas/clocked_rate_limiter.go (hunk 1 above)
Comment (prathyushpv) [reply common/quotas/clocked_rate_limiter.go]: I meant the caller to use Reserve() method first to get a reservation and then call Wait() on the reservation instead of calling Wait() on the ratelimiter.

File: common/quotas/clocked_rate_limiter.go [hunk 7]
```diff
@@ -151,3 +162,39 @@ func (l ClockedRateLimiter) SetBurstAt(t time.Time, newBurst int) {
...
+// rate limit AND there are many invalid tasks, we can make it a buffered channel instead.
+//
+// Our goal is to ensure that each token in our bucket is used every second, meaning the time between
+// taking and successfully using a token must be <= 1s. For this to be true, we must have:
+//
+//	time_to_recycle * number_of_recycles_per_second <= 1s
+//	time_to_recycle * probability_of_recycle * number_of_attempts_per_second <= 1s
+//
+// Therefore, it is also possible for this strategy to be inaccurate if the delay between taking and
+// successfully using a token is greater than one second.
```
Comment (dnr) [inline common/quotas/clocked_rate_limiter.go:189]: I know we talked about this but I don't understand it from this comment. There's nothing special about 1s... where does that number come from?

It seems like the longer the delay between taking and recycling, the more rate will be "wasted", and the shorter the delay the less will be wasted. But it's a continuous thing, isn't it?

File: common/quotas/clocked_rate_limiter_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dnr) [reply common/quotas/clocked_rate_limiter_test.go:187]: eventually is fine, or you can do it pretty easily with a channel too:
- make ctx a context with a 5s timeout
- each time you do `<-someChan` do it in a select with ctx.Done(), and assert failure if on context timeout

File: common/quotas/clocked_rate_limiter_test.go [hunk 8]
```diff
@@ -151,3 +151,81 @@ func TestClockedRateLimiter_Wait_DeadlineWouldExceed(t *testing.T) {
...
+		go func() {
+			assert.NoError(t, rl.Wait(ctx))
+			done <- struct{}{}
+		}()
+
+		// The waiting channel isn't enough by itself to confirm waiting, because
+		// the inner go func can be called without the rl.Wait() having finished.
+		// Use time.Sleep to allow rl.Wait() to complete before signalling `waiting`.
+		// 1-millisecond sleep succeeded >50,000x, whereas 1-nanosecond sleep deadlocks.
+		time.Sleep(time.Millisecond)
```
Comment (dnr) [inline common/quotas/clocked_rate_limiter_test.go:178]: I would use 10ms as a value for "wait for another goroutine to start and get blocked in something" just to be absolutely sure. remember flakiness in CI is going to be a lot worse than locally

File: common/quotas/clocked_rate_limiter_test.go (hunk 5 above)
Comment (dnr) [reply common/quotas/clocked_rate_limiter_test.go:168]: I agree the outer goroutine doesn't add anything here. the main goroutine just waits on waiting anyway

File: service/matching/matcher_test.go [hunk 9]
```diff
@@ -152,7 +152,7 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
 		task, err := t.childMatcher.Poll(ctx, &pollMetadata{})
 		cancel()
 		if err == nil && !task.isStarted() {
-			task.finish(nil)
+			task.finish(nil, true)
```
Comment (dnr) [inline service/matching/matcher_test.go:155]: not this line, but line 167:
there's going to be another merge conflict with #6535, btw

File: common/quotas/clocked_rate_limiter.go (hunk 1 above)
Comment (dnr) [reply common/quotas/clocked_rate_limiter.go]: I agree that this is overkill for now. We should just comment on RecycleToken that it recycles a single token and is only useful for the case where you took one.

File: common/quotas/clocked_rate_limiter.go [hunk 10]
```diff
@@ -131,12 +133,21 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
...
+	for {
+		select {
+		case <-ctx.Done():
+			reservation.Cancel()
+			return fmt.Errorf("%w: %v", ErrRateLimiterWaitInterrupted, ctx.Err())
+		case <-waitExpired:
+			return nil
+		case <-l.recycleCh:
+			if token > 1 {
+				break
```
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go:146]: My intention here is to make RecycleToken only unblock a waiter if the waiter is asking for 1 token, since I felt like that behavior would be easier to guarantee and test, and it's rare that we ask for more than one rate limit token in temporal in general, so this behavior covers most use cases. All the rate limit waits in matching are only using one token, so this is good enough there.

File: common/quotas/clocked_rate_limiter.go (hunk 7 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go:190]: hm, I thought that 1s comes from the fact that our rate limits are specified as RPS, so the rate limiter has a certain # of tokens to dole out each second. But I totally could have misunderstood -- I didn't read the innermost parts of the rate limit package that we import

File: common/quotas/clocked_rate_limiter.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go:146]: Looks like you read the other comment thread about this and agree as long as the behavior is commented, which it is! Resolving this comment for bookkeeping hope that's alright.

File: common/quotas/clocked_rate_limiter.go [hunk 11]
```diff
@@ -131,12 +133,21 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
...
+			}
+			reservation.Cancel()
```
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go:161]: done!

File: common/quotas/clocked_rate_limiter_test.go (hunk 5 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter_test.go:168]: I see, thank you! I was still thinking about the function without the sleep (but obviously the channel alone was not enough, so the sleep is doing the required work here)

File: common/quotas/clocked_rate_limiter_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (carlydf) [reply common/quotas/clocked_rate_limiter_test.go:187]: fixed, thanks for catching!

File: common/quotas/clocked_rate_limiter.go [hunk 12]
```diff
@@ -131,12 +133,34 @@ func (l ClockedRateLimiter) WaitN(ctx context.Context, token int) error {
...
+		case <-l.recycleCh:
+			if token > 1 {
+				break // recycling 1 token to a process requesting >1 tokens is a no-op
+			}
+
+			// Cancel() reverses the effects of this Reservation on the rate limit as much as possible,
+			// considering that other reservations may have already been made. Normally, Cancel() indicates
+			// that the reservation holder will not perform the reserved action, so it would make the most
+			// sense to cancel the reservation whose token was just recycled. However, we don't have access
+			// to the recycled reservation anymore, so instead we cancel the current reservation as a proxy.
```
Comment (dnr) [inline common/quotas/clocked_rate_limiter.go:153]: even if we did have access to it, Cancel() would be a no-op on that one since it's already past its time...
```suggestion
			// to the recycled reservation anymore, and even if we did, Cancel on a reservation that
			// has fully waited is a no-op, so instead we cancel the current reservation as a proxy.
```

File: common/quotas/clocked_rate_limiter.go (hunk 7 above)
Comment (dnr) [reply common/quotas/clocked_rate_limiter.go:203]: yeah but it doesn't add the tokens at the top of each second, it's continuous. so the units of the rate shouldn't matter.

File: common/quotas/clocked_rate_limiter_test.go [hunk 13]
```diff
@@ -151,3 +151,65 @@ func TestClockedRateLimiter_Wait_DeadlineWouldExceed(t *testing.T) {
...
+	t.Parallel()
+	ts := clock.NewEventTimeSource()
+	rl := quotas.NewClockedRateLimiter(rate.NewLimiter(1, 1), ts)
+	ctx := context.Background()
+
+	// take first token
+	assert.NoError(t, rl.Wait(ctx))
+
+	// wait for next token and report when success
+	asserted := false
```
Comment (dnr) [inline common/quotas/clocked_rate_limiter_test.go:167]: should be an atomic.Bool if you writing from one goroutine and reading from another. or use a channel.

File: common/quotas/clocked_rate_limiter_test.go [hunk 14]
```diff
@@ -151,3 +151,65 @@ func TestClockedRateLimiter_Wait_DeadlineWouldExceed(t *testing.T) {
...
+	go func() {
+		assert.NoError(t, rl.Wait(ctx))
+		asserted = true
+	}()
+	// wait for rl.Wait() to start and get to the select statement
+	time.Sleep(time.Second) // nolint
```
Comment (dnr) [inline common/quotas/clocked_rate_limiter_test.go:173]: This doesn't seem right.. at the start of the second there are no tokens, but after a whole second, there should be 1 (since the rate is 1/sec). so this Wait could return either naturally or due to the Recycle, and we can't tell the difference. It should more like sleep 10ms to let it get blocked, then recycle, then check that Wait returned within 10ms after that (20ms total), much less than 1s (the natural wakeup)

File: common/quotas/clocked_rate_limiter.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dnr) [reply common/quotas/clocked_rate_limiter.go]: yeah, my point was that we could make the receiving side logic correct, without worrying about the sending side yet. I admit it's a little silly since it's not tested. so either way is fine

File: common/quotas/clocked_rate_limiter_test.go (hunk 14 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter_test.go:173]: Yeah, the test wasn't passing in CI due to detecting a race, so I tried the 1s sleep just to see if it would help the test not have a race. But that didn't even work :/

File: common/quotas/clocked_rate_limiter.go (hunk 12 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter.go:153]: you're right, thanks for catching

File: common/quotas/clocked_rate_limiter_test.go (hunk 14 above)
Comment (carlydf) [reply common/quotas/clocked_rate_limiter_test.go]: reverted to 10ms
